src/digest.rs

```diff
@@ -448,8 +447,66 @@ pub extern fn SHA512_4(out: *mut u8, out_len: c::size_t,
     polyfill::slice::fill_from_slice(out, digest);
 }
 
+unsafe extern fn sha1_block_data_order(state: *mut u64, data: *const u8, num: c::size_t) {
+    sha1_process_blocks(
+        &mut *(state as *mut _),
+        core::slice::from_raw_parts(data as *const _, num)
+    )
+}
+
+type Sha1State = [u32; 5];
+type Sha1Block = [u32; 16];
+
+fn sha1_process_blocks(state: &mut Sha1State, blocks: &[Sha1Block]) {
+    #[inline] fn ch(x: u32, y: u32, z: u32) -> u32 { (x & y) | (!x & z) }
+    #[inline] fn parity(x: u32, y: u32, z: u32) -> u32 { x ^ y ^ z }
+    #[inline] fn maj(x: u32, y: u32, z: u32) -> u32 { (x & y) | (x & z) | (y & z) }
+
+    let mut w: [u32; 80] = [0; 80];
+    for block in blocks {
+        for t in 0..16 {
+            w[t] = u32::from_be(block[t])
+        }
+        for t in 16..80 {
+            w[t] = (w[t - 3] ^ w[t - 8] ^ w[t - 14] ^ w[t - 16]).rotate_left(1)
+        }
+
+        let mut a = state[0];
+        let mut b = state[1];
+        let mut c = state[2];
+        let mut d = state[3];
+        let mut e = state[4];
+
+
+        for t in 0..80 {
+            let (k, f) = match t {
+                0...19 => (0x5a827999, ch(b, c, d)),
+                20...39 => (0x6ed9eba1, parity(b, c, d)),
+                40...59 => (0x8f1bbcdc, maj(b, c, d)),
+                60...79 => (0xca62c1d6, parity(b, c, d)),
+                _ => unreachable!()
+            };
+            let tt = a.rotate_left(5)
+                .wrapping_add(f)
```

Comment:
I guess we should file a bug against the Rust stdlib about the missing `rotate_left` for `Wrapping`. We can define our own polyfill in `ring::polyfill` in the meantime; something like this:

```
// `core::num::Wrapping` doesn't support `rotate_left`. There is no usable
// trait for `rotate_left`, so this polyfill just hard-codes u32.
// https://github.com/rust-lang/rust/issues/XXXX
fn wrapping_rotate_left_u32(x: core::num::Wrapping<u32>, n: u32)
                            -> core::num::Wrapping<u32> {
   let core::num::Wrapping(x) = x;
   core::num::Wrapping(x.rotate_left(n)) 
}
```

This would allow us to use Wrapping<u32> for the time of every variable we do math on here, I think. It's ugly, but hopefully the ugliness is only temporary.